### PR TITLE
[Mypy] Add type check with mypy and fix storage

### LINKED ISSUE #1510: [Storage] Uploading list of files to bucket fails
Seems the following test failed, after this #1494, cc @infwinston:
```
_________________________________________________________________________________________________ TestStorageWithCredentials.test_list_source[StoreType.S3] __________________________________________________________________________________________________
[gw0] linux -- Python 3.10.6 /home/ubuntu/.conda/envs/sky/bin/python3.1
tests/test_smoke.py:1361: in test_list_source
    tmp_local_list_storage_obj.add_store(store_type)
sky/data/storage.py:658: in add_store
    self._sync_store(store)
sky/data/storage.py:725: in _sync_store
    os.path.join(self.source, '.git')):
../.conda/envs/sky/lib/python3.10/posixpath.py:76: in join
    a = os.fspath(a)
E   TypeError: expected str, bytes or os.PathLike object, not list
```

I believe the reason is that we did not take the `self.source` to be a list into account in the following place. https://github.com/skypilot-org/skypilot/blob/742c1dc82e62322d4e6d61c68b021b816f6110ad/sky/data/storage.py#L723-L728

### PR DESCRIPTION
**Pull request by Michaelvll**
Describe the changes in this PR:

Fixes #1510.

This PR also adds the mypy type check for a single file `sky/data/storage.py`. The files to be checked should be added incrementally. 

Tested (run the relevant ones):

- [x] `bash tests/run_smoke_tests.sh`

### CODE REVIEW

File: sky/data/storage.py [hunk 1]
```diff
@@ -125,7 +130,7 @@ class StoreMetadata:
         def __init__(self,
                      *,
                      name: str,
-                     source: str,
+                     source: Union[Path, List[Path], None],
```
Comment (romilbhardwaj) [inline sky/data/storage.py:133]: Should we define a variable `source_type=Union[Path, List[Path]` and use that everywhere? (Please feel free to use a better name for the variable)

File: sky/data/storage.py [hunk 2]
```diff
@@ -545,21 +539,25 @@ def _validate_local_source(local_source):
                         f'Supported paths: local, s3://, gs://. Got: {source}')
         return source, is_local_source
 
-    def _validate_storage_spec(self) -> None:
+    def _validate_storage_spec(
+        self,
+        name: Optional[str],
+        source: Union[Path, List[Path], None],
+    ) -> Tuple[str, Union[Path, List[Path], None]]:
```
Comment (romilbhardwaj) [inline sky/data/storage.py:546]: Any particular reason to have this method return values instead of updating `self.source` and `self.name` in place?

**Review by romilbhardwaj — COMMENTED** [review]
This is a great step towards catching bugs early, thanks for adding @Michaelvll! Left some comments

File: format.sh [hunk 3]
```diff
@@ -101,3 +121,7 @@ if ! git diff --quiet &>/dev/null; then
 
     exit 1
 fi
+
+# Run mypy
+echo 'SkyPilot MYPY:'
```
Comment (romilbhardwaj) [inline format.sh:126]: nit: `SkyPilot mypy:`

File: sky/data/storage.py [hunk 4]
```diff
@@ -308,7 +302,7 @@ def __init__(
             self,
             *,
             storage_name: Optional[str],
-            source: Optional[str],
+            source: Union[str, List[str], None],
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/data/storage.py]: Good point! Just switched to `Optional` with the defined `SourceType`

File: sky/data/storage.py [hunk 5]
```diff
@@ -449,8 +442,9 @@ def __init__(self,
                         self.add_store(StoreType.GCS)
 
     @staticmethod
-    def _validate_source(source: str, mode: StorageMode,
-                         sync_on_reconstruction: bool) -> Tuple[str, bool]:
+    def _validate_source(
+            source: Union[str, List[str]], mode: StorageMode,
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/data/storage.py]: Done. Thanks!

File: sky/data/storage.py (hunk 1 above)
Comment (Michaelvll) [reply sky/data/storage.py]: Good idea! Applied the changes.

File: sky/data/storage.py [hunk 6]
```diff
@@ -545,21 +539,25 @@ def _validate_local_source(local_source):
                         f'Supported paths: local, s3://, gs://. Got: {source}')
         return source, is_local_source
 
-    def _validate_storage_spec(self) -> None:
+    def _validate_storage_spec(
+        self,
+        name: Optional[str],
+        source: Union[Path, List[Path], None],
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/data/storage.py]: Fixed. Thanks!

File: sky/data/storage.py (hunk 2 above)
Comment (Michaelvll) [reply sky/data/storage.py]: It was a remnant for fixing the mypy error. Just reverted it. Thanks!

File: sky/data/storage.py [hunk 7]
```diff
@@ -390,15 +382,16 @@ def __init__(self,
         self.force_delete = False
 
         # Validate and correct inputs if necessary
-        self._validate_storage_spec()
+        self.name, self.source = self._validate_storage_spec(name, source)
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/data/storage.py]: Since we reverted the changes here. It might not be neccessary

**Review by romilbhardwaj — APPROVED** [review]
Awesome, lgtm!

**Comment by Michaelvll** [comment]
Thank you for the review @romilbhardwaj! Just passed the tests relate to storage. Merging.

**Pull request merged**